plugins/gpio/fu-plugin-gpio.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+static void
+fu_plugin_gpio_init(FuPlugin *plugin)
+{
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	fu_context_add_quirk_key(ctx, "GpioPrepare");
+	fu_context_add_quirk_key(ctx, "GpioCleanup");
+}
+
+static gboolean
+fu_plugin_gpio_create_node(FuPlugin *self, guint gpio_num, GError **error)
+{
+	g_autofree gchar *devname = g_strdup_printf("gpio%u", gpio_num);
+	g_autofree gchar *fn = g_build_filename("/sys/class/gpio", devname, NULL);
+	g_autofree gchar *gpio_number_str = NULL;
+	g_autoptr(FuIOChannel) io_channel = NULL;
+
+	/* already exists, so NOP */
+	if (g_file_test(fn, G_FILE_TEST_EXISTS))
+		return TRUE;
+
+	/* echo the device number as a string */
+	io_channel = fu_io_channel_new_file("/sys/class/gpio/export", error);
+	if (io_channel == NULL)
+		return FALSE;
+	gpio_number_str = g_strdup_printf("%u", gpio_num);
+	return fu_io_channel_write_raw(io_channel,
+				       (const guint8 *)gpio_number_str,
+				       strlen(gpio_number_str),
+				       1500,
+				       FU_IO_CHANNEL_FLAG_SINGLE_SHOT,
+				       error);
+}
+
+static gboolean
+fu_plugin_gpio_set_direction(FuPlugin *self, guint gpio_num, const gchar *direction, GError **error)
+{
+	g_autofree gchar *fn = NULL;
+	g_autofree gchar *devname = g_strdup_printf("gpio%u", gpio_num);
+	g_autoptr(FuIOChannel) io_channel = NULL;
+
+	g_return_val_if_fail(direction != NULL, FALSE);
+
+	fn = g_build_filename("/sys/class/gpio", devname, "direction", NULL);
```

Comment:
for future debugging and evaluating true need, I think it might be useful to read the current value and direction of the GPIO before changing it.

Reply:
I nuked this code moving to the chardev mode.